File: docs/examples/onlineplay_sql.py
"""Server.md의 SQLite/MariaDB 저장소 예제. 기본 서버는 이 모듈을 불러오지 않는다."""
import json
import sqlite3


class SqlOnlinePlayStorage:
    """서비스와 같은 동기 계약으로 기존 JSON 문서를 SQL 테이블에 저장한다."""

    def initialize(self):
        for table in ("puyow_accounts", "puyow_rooms"):
            self.query(f"CREATE TABLE IF NOT EXISTS {table} (id VARCHAR(20) PRIMARY KEY, payload TEXT NOT NULL)")

    def query(self, sql, parameters=()):
        # 요청마다 연결을 열어 HTTP 스레드와 관리 스레드의 연결 공유를 피한다.
        connection = self.connect()
        try:
            cursor = connection.cursor()
            cursor.execute(sql, parameters)
            rows = cursor.fetchall() if sql.startswith("SELECT") else []
            connection.commit()
            return rows
        except Exception as error:
            connection.rollback()
            # 방 저장 실패를 처리하는 기존 서비스의 OSError 계약에 맞춘다.
            raise OSError("SQL 저장소 쿼리 실패") from error
        finally:
            connection.close()

    def load_nickname_index(self):
        index = {}
        try:
            for account_id, payload in self.query("SELECT id, payload FROM puyow_accounts"):
                try:
                    account = json.loads(payload)
                    if isinstance(account, dict) and isinstance(account.get("nickname"), str):
                        index[account["nickname"]] = account_id
                except (TypeError, ValueError):
                    pass
        except OSError:
            pass
        return index

    def list_accounts(self):
        # 관리 화면의 계정 목록용이다. 손상된 행은 닉네임 색인과 같이 건너뛴다.
        accounts = []
        try:
            for (payload,) in self.query("SELECT payload FROM puyow_accounts"):
                try:
                    account = json.loads(payload)
                    if isinstance(account, dict) and isinstance(account.get("id"), str):
                        accounts.append(account)
                except (TypeError, ValueError):
                    pass
        except OSError:
            pass
        return accounts

    def load_account(self, account_id):
        try:
            rows = self.query("SELECT payload FROM puyow_accounts WHERE id = ?", (account_id.lower(),))
            value = json.loads(rows[0][0]) if rows else None
            return value if isinstance(value, dict) else None
        except (OSError, TypeError, ValueError):
            return None

    def save_account(self, account):
        self.save("puyow_accounts", account)

    def save_room(self, snapshot):
        self.save("puyow_rooms", snapshot)

    def save(self, table, value):
        # 테이블명은 내부 상수만 사용하고 계정 값은 SQL에 이어 붙이지 않는다.
        suffix = ("ON DUPLICATE KEY UPDATE payload = VALUES(payload)" if self.dialect == "mariadb"
                  else "ON CONFLICT(id) DO UPDATE SET payload = excluded.payload")
        self.query(f"INSERT INTO {table} (id, payload) VALUES (?, ?) {suffix}",
                   (value["id"].lower(), json.dumps(value, ensure_ascii=False)))

    def remove_room(self, room_id):
        try:
            self.query("DELETE FROM puyow_rooms WHERE id = ?", (room_id.lower(),))
        except OSError:
            pass

    def clear_rooms(self):
        try:
            self.query("DELETE FROM puyow_rooms")
        except OSError:
            pass


class SqliteOnlinePlayStorage(SqlOnlinePlayStorage):
    def __init__(self, filename):
        # 메서드별 연결을 쓰므로 :memory: 대신 실제 파일 경로를 사용한다.
        self.filename = filename
        self.dialect = "sqlite"

    def connect(self):
        try:
            return sqlite3.connect(self.filename)
        except sqlite3.Error as error:
            raise OSError("SQLite 연결 실패") from error


class MariaDbOnlinePlayStorage(SqlOnlinePlayStorage):
    def __init__(self, config):
        self.config = config
        self.dialect = "mariadb"

    def connect(self):
        import mariadb
        try:
            return mariadb.connect(**self.config)
        except mariadb.Error as error:
            raise OSError("MariaDB 연결 실패") from error

File: docs/examples/onlineplay_sql.js
/** Server.md의 동기식 SQL 저장소 예제. 기본 서버는 이 파일을 사용하지 않는다. */
const path = require('node:path');
const { spawnSync } = require('node:child_process');

/** 테이블에 기존 JSON을 그대로 저장하여 필드와 시간 단위를 유지한다. */
class SqlOnlinePlayStorage {
    /** 스키마 준비와 방 초기화는 분리한다. 계정은 초기화할 때 삭제하지 않는다. */
    initialize() {
        for (const table of ['puyow_accounts', 'puyow_rooms']) {
            this.query(`CREATE TABLE IF NOT EXISTS ${table} (id VARCHAR(20) PRIMARY KEY, payload TEXT NOT NULL)`);
        }
    }

    /** 서비스가 닉네임 대소문자 구분을 계속 담당한다. */
    loadNicknameIndex() {
        const index = new Map();
        try {
            for (const row of this.query('SELECT id, payload FROM puyow_accounts')) {
                try {
                    const account = JSON.parse(row.payload);
                    if (typeof account?.nickname === 'string') index.set(account.nickname, row.id);
                } catch { /* 손상된 행은 기존 파일 색인과 같이 건너뛴다. */ }
            }
        } catch { /* 기존 파일 조회 실패 정책을 유지하는 학습용 예제다. */ }
        return index;
    }

    /** 관리 화면의 계정 목록용이다. 손상된 행은 닉네임 색인과 같이 건너뛴다. */
    listAccounts() {
        const accounts = [];
        try {
            for (const row of this.query('SELECT payload FROM puyow_accounts')) {
                try {
                    const account = JSON.parse(row.payload);
                    if (typeof account?.id === 'string') accounts.push(account);
                } catch { /* 손상된 행은 건너뛴다. */ }
            }
        } catch { /* 기존 파일 조회 실패 정책을 유지하는 학습용 예제다. */ }
        return accounts;
    }

    loadAccount(id) {
        try {
            const rows = this.query('SELECT payload FROM puyow_accounts WHERE id = ?', [id.toLowerCase()]);
            return rows.length ? JSON.parse(rows[0].payload) : null;
        } catch { return null; }
    }

    saveAccount(account) { this.save('puyow_accounts', account); }
    saveRoom(snapshot) { this.save('puyow_rooms', snapshot); }

    /** 테이블명은 내부 상수만 사용하고 값은 SQL 매개변수로 전달한다. */
    save(table, value) {
        const suffix = this.dialect === 'mariadb'
            ? 'ON DUPLICATE KEY UPDATE payload = VALUES(payload)'
            : 'ON CONFLICT(id) DO UPDATE SET payload = excluded.payload';
        this.query(`INSERT INTO ${table} (id, payload) VALUES (?, ?) ${suffix}`, [value.id.toLowerCase(), JSON.stringify(value)]);
    }

    removeRoom(id) {
        try { this.query('DELETE FROM puyow_rooms WHERE id = ?', [id.toLowerCase()]); }
        catch { /* 기존 삭제 실패 정책을 유지한다. */ }
    }

    clearRooms() {
        try { this.query('DELETE FROM puyow_rooms'); }
        catch { /* 메모리의 방 상태가 정본이다. */ }
    }
}

/** Node 22.12에서는 --experimental-sqlite 옵션이 필요하다. */
class SqliteOnlinePlayStorage extends SqlOnlinePlayStorage {
    constructor(filename) {
        super();
        this.filename = filename;
        this.dialect = 'sqlite';
    }

    initialize() {
        const { DatabaseSync } = require('node:sqlite');
        this.database = new DatabaseSync(this.filename);
        super.initialize();
    }

    query(sql, parameters = []) {
        const statement = this.database.prepare(sql);
        if (sql.startsWith('SELECT')) return statement.all(...parameters);
        statement.run(...parameters);
        return [];
    }

    close() { this.database?.close(); }
}

/** 동기식 서비스와 비동기 DB 드라이버 사이의 학습용 별도 프로세스 경계다. */
class MariaDbOnlinePlayStorage extends SqlOnlinePlayStorage {
    constructor(config) {
        super();
        this.config = config;
        this.dialect = 'mariadb';
    }

    query(sql, parameters = []) {
        // 비밀번호를 명령행에 넣지 않고 표준 입력으로 보낸다. 셸은 실행하지 않는다.
        const result = spawnSync(process.execPath, [path.join(__dirname, 'onlineplay_mariadb_query.js')], {
            input: JSON.stringify({ config: this.config, sql, parameters }),
            encoding: 'utf8', windowsHide: true, timeout: 15000, maxBuffer: 8 * 1024 * 1024
        });
        if (result.error || result.status !== 0) throw new Error('MariaDB 예제 쿼리를 실행하지 못했습니다.');
        return JSON.parse(result.stdout);
    }
}

module.exports = { SqliteOnlinePlayStorage, MariaDbOnlinePlayStorage };
